File: web/math/mat3.ts
import Vec2 from "./vec2";

export default class Mat3 {
  constructor(
    readonly m11: number,
    readonly m21: number,
    readonly m31: number,
    readonly m12: number,
    readonly m22: number,
    readonly m32: number,
    readonly m13: number,
    readonly m23: number,
    readonly m33: number
  ) {}

  determinant(): number {
    return (
      this.m11 * (this.m22 * this.m33 - this.m32 * this.m23) -
      this.m12 * (this.m12 * this.m33 - this.m32 * this.m13) +
      this.m13 * (this.m12 * this.m23 - this.m22 * this.m13)
    );
  }

  inverse(): Mat3 {
    const x = 1 / this.determinant();

    return new Mat3(
      (this.m22 * this.m33 - this.m32 * this.m23) * x,
      (this.m31 * this.m23 - this.m21 * this.m33) * x,
      (this.m21 * this.m32 - this.m31 * this.m22) * x,

      (this.m32 * this.m13 - this.m12 * this.m33) * x,
      (this.m11 * this.m33 - this.m31 * this.m13) * x,
      (this.m31 * this.m12 - this.m11 * this.m32) * x,

      (this.m12 * this.m23 - this.m22 * this.m13) * x,
      (this.m21 * this.m13 - this.m11 * this.m23) * x,
      (this.m11 * this.m22 - this.m21 * this.m12) * x
    );
  }

  apply(v: Vec2): Vec2 {
    return new Vec2(
      this.m11 * v.x + this.m21 * v.y + this.m31,
      this.m12 * v.x + this.m22 * v.y + this.m32
    );
  }

  static identity = new Mat3(1, 0, 0, 0, 1, 0, 0, 0, 1);

  static mul(left: Mat3, right: Mat3): Mat3 {
    return new Mat3(
      left.m11 * right.m11 + left.m21 * right.m12 + left.m31 * right.m13,
      left.m11 * right.m21 + left.m21 * right.m22 + left.m31 * right.m23,
      left.m11 * right.m31 + left.m21 * right.m32 + left.m31 * right.m33,

      left.m12 * right.m11 + left.m22 * right.m12 + left.m32 * right.m13,
      left.m12 * right.m21 + left.m22 * right.m22 + left.m32 * right.m23,
      left.m12 * right.m31 + left.m22 * right.m32 + left.m32 * right.m33,

      left.m13 * right.m11 + left.m23 * right.m12 + left.m33 * right.m13,
      left.m13 * right.m21 + left.m23 * right.m22 + left.m33 * right.m23,
      left.m13 * right.m31 + left.m23 * right.m32 + left.m33 * right.m33
    );
  }

  static translation(vector: Vec2): Mat3 {
    return new Mat3(1, 0, vector.x, 0, 1, vector.y, 0, 0, 1);
  }

  static scale(vector: Vec2): Mat3 {
    return new Mat3(vector.x, 0, 0, 0, vector.y, 0, 0, 0, 1);
  }
}

File: rust/src/util/mat3.rs
use crate::util::vector::{vec2, vec3, Vec2, Vec3};
use std::ops::{Mul, MulAssign};

#[derive(Copy, Clone, PartialEq)]
pub struct Mat3 {
    data: [[f64; 3]; 3],
}

impl Mat3 {
    pub fn new(data: [[f64; 3]; 3]) -> Self {
        Self { data }
    }

    pub fn identity() -> Self {
        Self::new([[1.0, 0.0, 0.0], [0.0, 1.0, 0.0], [0.0, 0.0, 1.0]])
    }

    pub fn scale(vector: Vec2) -> Self {
        Self::new([
            [vector.x(), 0.0, 0.0],
            [0.0, vector.y(), 0.0],
            [0.0, 0.0, 1.0],
        ])
    }

    pub fn translation(vector: Vec2) -> Self {
        Self::new([
            [1.0, 0.0, vector.x()],
            [0.0, 1.0, vector.y()],
            [0.0, 0.0, 1.0],
        ])
    }

    pub fn rotation(radians: f64) -> Self {
        let (sin, cos) = radians.sin_cos();

        Self::new([[cos, -sin, 0.0], [sin, cos, 0.0], [0.0, 0.0, 1.0]])
    }

    pub fn transform(translation: Vec2, scale: Vec2) -> Self {
        Self::translation(translation) * Self::scale(scale)
    }

    pub fn determinant(&self) -> f64 {
        self.m11() * (self.m22() * self.m33() - self.m32() * self.m23())
            - self.m12() * (self.m12() * self.m33() - self.m32() * self.m13())
            + self.m13() * (self.m12() * self.m23() - self.m22() * self.m13())
    }

    pub fn inverse(&self) -> Self {
        let x = 1.0 / self.determinant();

        return Self::new([
            [
                (self.m22() * self.m33() - self.m32() * self.m23()) * x,
                (self.m31() * self.m23() - self.m21() * self.m33()) * x,
                (self.m21() * self.m32() - self.m31() * self.m22()) * x,
            ],
            [
                (self.m32() * self.m13() - self.m12() * self.m33()) * x,
                (self.m11() * self.m33() - self.m31() * self.m13()) * x,
                (self.m31() * self.m12() - self.m11() * self.m32()) * x,
            ],
            [
                (self.m12() * self.m23() - self.m22() * self.m13()) * x,
                (self.m21() * self.m13() - self.m11() * self.m23()) * x,
                (self.m11() * self.m22() - self.m21() * self.m12()) * x,
            ],
        ]);
    }

    fn at(&self, x: usize, y: usize) -> f64 {
        self.data[y][x]
    }

    pub fn m11(&self) -> f64 {
        self.at(0, 0)
    }

    pub fn m21(&self) -> f64 {
        self.at(1, 0)
    }

    pub fn m31(&self) -> f64 {
        self.at(2, 0)
    }

    pub fn m12(&self) -> f64 {
        self.at(0, 1)
    }

    pub fn m22(&self) -> f64 {
        self.at(1, 1)
    }

    pub fn m32(&self) -> f64 {
        self.at(2, 1)
    }

    pub fn m13(&self) -> f64 {
        self.at(0, 2)
    }

    pub fn m23(&self) -> f64 {
        self.at(1, 2)
    }

    pub fn m33(&self) -> f64 {
        self.at(2, 2)
    }
}

impl Mul for Mat3 {
    type Output = Self;

    fn mul(self, rhs: Self) -> Self {
        Self::new([
            [
                self.m11() * rhs.m11() + self.m21() * rhs.m12() + self.m31() * rhs.m13(),
                self.m11() * rhs.m21() + self.m21() * rhs.m22() + self.m31() * rhs.m23(),
                self.m11() * rhs.m31() + self.m21() * rhs.m32() + self.m31() * rhs.m33(),
            ],
            [
                self.m12() * rhs.m11() + self.m22() * rhs.m12() + self.m32() * rhs.m13(),
                self.m12() * rhs.m21() + self.m22() * rhs.m22() + self.m32() * rhs.m23(),
                self.m12() * rhs.m31() + self.m22() * rhs.m32() + self.m32() * rhs.m33(),
            ],
            [
                self.m13() * rhs.m11() + self.m23() * rhs.m12() + self.m33() * rhs.m13(),
                self.m13() * rhs.m21() + self.m23() * rhs.m22() + self.m33() * rhs.m23(),
                self.m13() * rhs.m31() + self.m23() * rhs.m32() + self.m33() * rhs.m33(),
            ],
        ])
    }
}

impl MulAssign for Mat3 {
    fn mul_assign(&mut self, rhs: Self) {
        *self = *self * rhs;
    }
}

impl Mul<Vec2> for Mat3 {
    type Output = Vec2;

    fn mul(self, rhs: Vec2) -> Vec2 {
        vec2(
            self.m11() * rhs.x() + self.m21() * rhs.y() + self.m31(),
            self.m12() * rhs.x() + self.m22() * rhs.y() + self.m32(),
        )
    }
}

impl Mul<Vec3> for Mat3 {
    type Output = Vec3;

    fn mul(self, rhs: Vec3) -> Vec3 {
        vec3(
            self.m11() * rhs.x() + self.m21() * rhs.y() + self.m31() * rhs.z(),
            self.m12() * rhs.x() + self.m22() * rhs.y() + self.m32() * rhs.z(),
            self.m13() * rhs.x() + self.m23() * rhs.y() + self.m33() * rhs.z(),
        )
    }
}
